File: ctci/tmp/chapter_3_stacks_and_queues/cpp/3.2-stack_min.cpp
#include <iostream>
#include <limits>
#include <stack>

using namespace std;

class StackWithMin
{
public:
    stack<int> s1;
    /* New stack with minimums. */
    stack<int> s2;

    StackWithMin(){};
    void push(int value);
    int pop();
    int min();
};

/**
 * Pushes an element to the top of the stack.
 */
void StackWithMin::push(int value)
{
    if (value < this->min())
        s2.push(value);
    s1.push(value);
}

/**
 * Pops an element off the top of the stack.
 */
int StackWithMin::pop()
{
    int value = s1.top();
    if (value == this->min())
        /* If min is getting popped, remove from second list also. */
        s2.pop();
    s1.pop();
    return value;
}

/**
 * Return the minimum value of the stack.
 */
int StackWithMin::min()
{
    if (s2.empty())
        return INT32_MAX;
    return s2.top();
}

int main()
{
    StackWithMin swm;
    swm.push(1);
    swm.push(-5);
    swm.push(10);
    cout << swm.min() << endl; // -5
    swm.push(-10);
    cout << swm.min() << endl; // -10
    cout << swm.pop() << endl; // -10
    cout << swm.min() << endl; // -5

    return 0;
}

File: ctci/tmp/chapter_3_stacks_and_queues/python/3.2-stack_min.py
Stack = __import__('stack').Stack


class StackWithMin(Stack):
    def __init__(self):
        """
        Stack with minimum constructor.
        """
        super().__init__()
        # new stack with minimums
        self.s2 = Stack()

    def push(self, value: int):
        """
        Pushes an element to the top of the stack.
        """
        if value < self.min():
            self.s2.push(value)
        super().push(value)

    def pop(self) -> int:
        """
        Pops an element off the top of the stack.
        """
        # If min is getting popped, remove from second list also.
        tmp = super().pop()
        if tmp == self.min():
            self.s2.pop()
        return tmp

    def min(self) -> int:
        """
        Return the minimum value of the stack.
        """
        if self.s2.is_empty():
            return float('inf')
        return self.s2.peek()


if __name__ == 'main':
    swm = StackWithMin()
    print(swm.__dict__)
    # {'stack': [], 'length': 0, 's2': <stack.Stack objek object at 0x7f6a0681a390>}
    for i in [4, -1, 2, 6, 2, -5]:
        swm.push(i)
    print(swm.min())  # -5
    print(swm.stack)  # [4, -1, 2, 6, 2, -5]
    print(swm.s2.stack)  # [4, -1, -5]
    print(swm.pop())  # -5
    print(swm.stack)  # [4, -1, 2, 6, 2]
    print(swm.s2.stack)  # [4, -1]
